inst/include/dust/r/dust.hpp

```diff
@@ -152,6 +152,11 @@ template <typename T, typename std::enable_if<std::is_same<size_t, typename T::t
 void dust_initialise(T *obj, bool reset_step_size) {
 }
 
+template <typename T, typename std::enable_if<std::is_same<double, typename T::time_type>::value, int>::type = 0>
```

Comment:
this syntax is :open_mouth: I am reading the docs and still struggling to grasp what this means

Reply:
yeah, C++ is terrible, I know. This is an example of [SFINAE](https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error) which is the metaprogramming technique that C++ uses to get generic functions that get resolved at compile time. Basically we want two forms of this (and other) functions where the correct version gets used depending on some trait-like thing in a type

```
// use this version where T::time_type is size_t  - i.e., discrete time models
template <typename T, typename std::enable_if<std::is_same<size_t, typename T::time_type>::value, int>::type = 0>
void dust_initialise(T *obj, bool reset_step_size) {
}

// use this version where T::time_type is double  - i.e., ode models
template <typename T, typename std::enable_if<std::is_same<double, typename T::time_type>::value, int>::type = 0>
void dust_initialise(T *obj, bool reset_step_size) {
  // It does not matter here that a discrete time model will not have an initialise method, as this version of the function never gets instantiated for a that sort of model
  obj->initialise(reset_step_size);
}
```

The parts here are `enable_if` which is does the test - if it fails then no suitable type can be found for the second template parameter causing that template to be passed over for the next one